sys/net/routing/ng_rpl/ng_rpl_control_messages.c

```diff
@@ -0,0 +1,577 @@
+    pos = (uint8_t *) dio;
+    dio->instance_id = dodag->instance->id;
+    dio->version_number = dodag->version;
+    dio->rank = byteorder_htons(dodag->my_rank);
+    dio->g_mop_prf = (dodag->grounded << NG_RPL_GROUNDED_SHIFT) |
+        (dodag->instance->mop << NG_RPL_MOP_SHIFT) | dodag->prf;
+    dio->dtsn = dodag->dtsn;
+    dio->flags = 0;
+    dio->reserved = 0;
+    dio->dodag_id = dodag->dodag_id;
+
+    pos += sizeof(*dio);
+
+    if ((dodag->dodag_conf_counter % 3) == 0) {
+        dodag_conf = (ng_rpl_opt_dodag_conf_t *) pos;
+        dodag_conf->type = NG_RPL_OPT_DODAG_CONF;
+        dodag_conf->length = NG_RPL_OPT_DODAG_CONF_LEN;
+        dodag_conf->flags_a_pcs = 0;
+        dodag_conf->dio_int_doubl = dodag->dio_interval_doubl;
+        dodag_conf->dio_int_min = dodag->dio_min;
+        dodag_conf->dio_redun = dodag->dio_redun;
+        dodag_conf->max_rank_inc = byteorder_htons(dodag->instance->max_rank_inc);
+        dodag_conf->min_hop_rank_inc = byteorder_htons(dodag->instance->min_hop_rank_inc);
+        dodag_conf->ocp = byteorder_htons(dodag->instance->of->ocp);
+        dodag_conf->reserved = 0;
+        dodag_conf->default_lifetime = dodag->default_lifetime;
+        dodag_conf->lifetime_unit = byteorder_htons(dodag->lifetime_unit);
+        pos += sizeof(*dodag_conf);
+    }
+
+    dodag->dodag_conf_counter++;
+    _ng_rpl_send(pkt, NULL, destination);
+}
+
+void ng_rpl_send_DIS(ng_rpl_dodag_t *dodag, ng_ipv6_addr_t *destination)
+{
+    (void) dodag;
+    ng_pktsnip_t *pkt;
+    ng_icmpv6_hdr_t *icmp;
+    ng_rpl_dis_t *dis;
+    int size = sizeof(ng_icmpv6_hdr_t) + sizeof(ng_rpl_dis_t) + 4;
+
+    if ((pkt = ng_icmpv6_build(NULL, NG_ICMPV6_RPL_CTRL, NG_RPL_ICMPV6_CODE_DIS, size)) == NULL) {
+        DEBUG("RPL: Send DIS - no space left in packet buffer\n");
+        return;
+    }
+
+    icmp = (ng_icmpv6_hdr_t *)pkt->data;
+    dis = (ng_rpl_dis_t *)(icmp + 1);
+    dis->flags = 0;
+    dis->reserved = 0;;
+    memset((dis + 1), 0, 4);
+
+    _ng_rpl_send(pkt, NULL, destination);
+}
+
+void ng_rpl_recv_DIS(ng_rpl_dis_t *dis, ng_ipv6_addr_t *src, ng_ipv6_addr_t *dst, uint16_t len)
+{
+    /* TODO handle Solicited Information Option */
+    (void) dis;
+    (void) len;
+
+    if (ng_ipv6_addr_is_multicast(dst)) {
+        for (uint8_t i = 0; i < NG_RPL_DODAGS_NUMOF; ++i) {
+            if (ng_rpl_dodags[i].state != 0) {
+                trickle_reset_timer(&ng_rpl_dodags[i].trickle);
+            }
+        }
+    }
+    else {
+        for (uint8_t i = 0; i < NG_RPL_DODAGS_NUMOF; ++i) {
+            if (ng_rpl_dodags[i].state != 0) {
+                ng_rpl_dodags[i].dodag_conf_counter = 0;
+                ng_rpl_send_DIO(&ng_rpl_dodags[i], src);
+            }
+        }
+    }
+}
+
+void _parse_options(ng_rpl_dodag_t *dodag, ng_rpl_opt_t *opt, uint16_t len, ng_ipv6_addr_t *src)
+{
+    uint16_t l = 0;
+    ng_rpl_opt_target_t *first_target = NULL;
+    while(l < len) {
+        switch(opt->type) {
```

Comment:
Do you think this is really necessary? IMHO this kind of validation is unneeded in this situation, because the while loop will anyways terminate when `l >= len`.
Besides, the proposed check will not find options with bad option lengths if it still fits into `len - l`.